source/windows/mutex.c

```diff
@@ -17,6 +17,8 @@
 
 int aws_mutex_init(struct aws_mutex *mutex, struct aws_allocator *allocator) {
     mutex->allocator = allocator;
+    /* A more efficient variant of InitializeSRWLock.
+       See https://msdn/microsoft.com/en-us/library/windows/desktop/ms683483(v=vs.85).aspx */
```

Comment:
Link is typoed. Also the linked URL does not explain why setting `.Ptr = 0` is a legal way to initialize the lock.

Why aren't we just calling InitializeSRWLock anyway?

Reply:
"To initialize the structure statically, assign the constant SRWLOCK_INIT to the structure variable." isn't an explanation?

"Why aren't we just calling InitializeSRWLock anyway?"

Because of this language:
"The InitializeSRWLock function is used to initialize a SRW lock dynamically."

What does this mean? Does it do an allocation? It's marked as an output parameter.

Reply:
It cannot do an allocation because it is documented that SRW locks are not required to be cleaned up or deallocated. I think you're overthinking this - SRW locks don't care what kind of memory they're used in; you can use InitializeSRWLock on a static variable if you want.

In any case, if you're doing something that isn't in the docs - i.e. setting a random member (which is not documented) to a constant zero (also not documented as a valid value for that undocumented member), you need to have a really convincing argument for why this is 1) permitted and 2) better than the documented methods. You can't just link a MSDN page which says to do something different from what you're actually doing.

Reply:
I'm doing exactly what the page says to do:

lock = SRWLOCK_INIT;

is literally identical to what I'm doing.

The macro replacement of SRWLOCK_INIT is {0}

I'm fine using the InitializeCall, but the attempt here is to follow the documentation, not the opposite.

Reply:
To be clear, the documentation says to use `SRWLOCK_INIT` for static initialization, which is a macro which could be redefined in the future. For example, it could be redefined to be `{1}` on a new platform, or the name or type of the member could change. While existing code on x86 or x86_64 windows must remain ABI compatible, code built on newer SDKs might break, or code on (insert fancy new mobile architecture) might break as well.

When something says `static initialization` in this context, it means initializing something at compile time (vs initializing it at runtime). The documentation doesn't say anything that would imply that `InitializeSRWLock` would be unacceptable in any context.

Reply:
like I said, I'm fine using the call.

It's not the words "static initialization" that is confusing, it's the words: "The InitializeSRWLock function is used to initialize a SRW lock dynamically" that is ambiguous. It implies that it wants to do an allocation. 

Anyways, let's close this thread, InitializeSRWLock is fine.